File: backend/src/compatibility.rs
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_identical_big_five_results() {
        let r1 = serde_json::json!({"scores": {"O": 70.0, "C": 60.0, "E": 50.0, "A": 80.0, "N": 30.0}});
        let r2 = serde_json::json!({"scores": {"O": 70.0, "C": 60.0, "E": 50.0, "A": 80.0, "N": 30.0}});
        let result = compare_results(&[r1, r2], "big_five", "en");
        let score = result["compatibility_score"].as_f64().unwrap();
        assert!(score >= 90.0, "identical results should have high compatibility, got {:.1}", score);
    }

    #[test]
    fn test_opposite_big_five_results() {
        let r1 = serde_json::json!({"scores": {"O": 90.0, "C": 90.0, "E": 90.0, "A": 90.0, "N": 10.0}});
        let r2 = serde_json::json!({"scores": {"O": 10.0, "C": 10.0, "E": 10.0, "A": 10.0, "N": 90.0}});
        let result = compare_results(&[r1, r2], "big_five", "en");
        let score = result["compatibility_score"].as_f64().unwrap();
        assert!(score < 60.0, "opposite results should have low compatibility, got {:.1}", score);
    }

    #[test]
    fn test_identical_mbti_results() {
        let r1 = serde_json::json!({"type_code": "INTJ", "scores": {"EI": 75.0, "SN": 20.0, "TF": 70.0, "JP": 20.0}});
        let r2 = serde_json::json!({"type_code": "INTJ", "scores": {"EI": 75.0, "SN": 20.0, "TF": 70.0, "JP": 20.0}});
        let result = compare_results(&[r1, r2], "mbti", "en");
        let score = result["compatibility_score"].as_f64().unwrap();
        assert!(score >= 85.0, "identical MBTI should have high compatibility, got {:.1}", score);
        assert_eq!(result["match_category"].as_str().unwrap(), "identical");
    }

    #[test]
    fn test_opposite_mbti_results() {
        let r1 = serde_json::json!({"type_code": "INTJ", "scores": {"EI": 75.0, "SN": 75.0, "TF": 75.0, "JP": 75.0}});
        let r2 = serde_json::json!({"type_code": "ESFP", "scores": {"EI": 25.0, "SN": 25.0, "TF": 25.0, "JP": 25.0}});
        let result = compare_results(&[r1, r2], "mbti", "en");
        let score = result["compatibility_score"].as_f64().unwrap();
        assert!(score <= 60.0, "opposite MBTI should have lower compatibility, got {:.1}", score);
    }

    #[test]
    fn test_unsupported_test_type() {
        let r1 = serde_json::json!({"scores": {"1": 50.0}});
        let r2 = serde_json::json!({"scores": {"1": 50.0}});
        let result = compare_results(&[r1, r2], "enneagram", "en");
        assert!(result.get("error").and_then(|v| v.as_bool()).unwrap_or(false),
                "enneagram comparison should return error");
    }

    #[test]
    fn test_big_five_has_factor_details() {
        let r1 = serde_json::json!({"scores": {"O": 50.0, "C": 50.0, "E": 50.0, "A": 50.0, "N": 50.0}});
        let r2 = serde_json::json!({"scores": {"O": 50.0, "C": 50.0, "E": 50.0, "A": 50.0, "N": 50.0}});
        let result = compare_results(&[r1, r2], "big_five", "en");
        let factors = result["factors"].as_object().unwrap();
        for &f in &["O", "C", "E", "A", "N"] {
            assert!(factors.contains_key(f), "missing factor {}", f);
            assert!(factors[f].get("score1").is_some(), "factor {} missing score1", f);
            assert!(factors[f].get("score2").is_some(), "factor {} missing score2", f);
            assert!(factors[f].get("difference").is_some(), "factor {} missing difference", f);
        }
    }

    #[test]
    fn test_compatibility_spanish() {
        let r1 = serde_json::json!({"scores": {"O": 70.0, "C": 60.0, "E": 50.0, "A": 80.0, "N": 30.0}});
        let r2 = serde_json::json!({"scores": {"O": 65.0, "C": 55.0, "E": 45.0, "A": 75.0, "N": 35.0}});
        let result = compare_results(&[r1, r2], "big_five", "es");
        assert!(result["description"].as_str().unwrap_or("").len() > 0);
    }
}

const FACTOR_NAMES_ES: [(&str, &str); 5] = [
    ("O", "Apertura"),
    ("C", "Responsabilidad"),
    ("E", "Extraversión"),
    ("A", "Amabilidad"),
    ("N", "Neuroticismo"),
];

const FACTOR_NAMES_EN: [(&str, &str); 5] = [
    ("O", "Openness"),
    ("C", "Conscientiousness"),
    ("E", "Extraversion"),
    ("A", "Agreeableness"),
    ("N", "Neuroticism"),
];

fn factor_names(lang: &str) -> &[(&str, &str); 5] {
    if lang == "en" {
        &FACTOR_NAMES_EN
    } else {
        &FACTOR_NAMES_ES
    }
}

fn clamped_diff(a: f64, b: f64) -> f64 {
    ((a - b).abs() * 10.0).round() / 10.0
}

fn clamp_score(v: f64) -> f64 {
    v.max(0.0).min(100.0)
}

pub fn compare_results(
    results: &[serde_json::Value],
    test_type: &str,
    lang: &str,
) -> serde_json::Value {
    match test_type {
        "big_five" => compare_big_five(results, lang),
        "mbti" => compare_mbti(results, lang),
        _ => serde_json::json!({
            "error": true,
            "detail": format!("Comparison not supported for test type '{}'", test_type),
        }),
    }
}

fn get_score(result: &serde_json::Value, factor: &str) -> f64 {
    result
        .get("scores")
        .and_then(|s| s.get(factor))
        .and_then(|v| v.as_f64())
        .unwrap_or(50.0)
}

fn compare_big_five(results: &[serde_json::Value], lang: &str) -> serde_json::Value {
    let names = factor_names(lang);
    let factors = ["O", "C", "E", "A", "N"];

    let r1 = results.first();
    let r2 = results.get(1);

    let mut total_diff = 0.0;
    let mut factor_details = serde_json::Map::new();

    for (i, f) in factors.iter().enumerate() {
        let s1 = r1.map(|r| get_score(r, f)).unwrap_or(50.0);
        let s2 = r2.map(|r| get_score(r, f)).unwrap_or(50.0);
        let diff = clamped_diff(s1, s2);
        total_diff += diff;

        let (_, label) = names[i];
        let both_high = s1 >= 60.0 && s2 >= 60.0;
        let both_low = s1 <= 40.0 && s2 <= 40.0;
        let desc = if both_high {
            if lang == "en" {
                format!("Both high in {}: you share this trait strongly.", label)
            } else {
                format!("Ambos altos en {}: comparten este rasgo fuertemente.", label)
            }
        } else if both_low {
            if lang == "en" {
                format!("Both low in {}: you share this tendency.", label)
            } else {
                format!("Ambos bajos en {}: comparten esta tendencia.", label)
            }
        } else if diff <= 15.0 {
            if lang == "en" {
                format!("Similar in {} (diff: {:.0}%).", label, diff)
            } else {
                format!("Similares en {} (dif: {:.0}%).", label, diff)
            }
        } else {
            if lang == "en" {
                format!(
                    "Complementary in {}: you balance each other (diff: {:.0}%).",
                    label, diff
                )
            } else {
                format!(
                    "Complementarios en {}: se equilibran mutuamente (dif: {:.0}%).",
                    label, diff
                )
            }
        };

        factor_details.insert(
            f.to_string(),
            serde_json::json!({
                "score1": (s1 * 10.0).round() / 10.0,
                "score2": (s2 * 10.0).round() / 10.0,
                "difference": diff,
                "description": desc,
            }),
        );
    }

    let avg_diff = total_diff / factors.len() as f64;
    let compatibility = clamp_score(100.0 - avg_diff);

    let summary = if lang == "en" {
        format!(
            "Compatibility: {:.0}% — {}",
            compatibility,
            if compatibility >= 80.0 {
                "Very compatible! You share similar traits."
            } else if compatibility >= 60.0 {
                "Good compatibility with some complementary differences."
            } else {
                "Opposite traits create a challenging dynamic."
            }
        )
    } else {
        format!(
            "Compatibilidad: {:.0}% — {}",
            compatibility,
            if compatibility >= 80.0 {
                "¡Muy compatibles! Comparten rasgos similares."
            } else if compatibility >= 60.0 {
                "Buena compatibilidad con diferencias complementarias."
            } else {
                "Rasgos opuestos crean una dinámica desafiante."
            }
        )
    };

    serde_json::json!({
        "test_type": "big_five",
        "compatibility_score": compatibility,
        "description": summary,
        "factors": factor_details,
    })
}

fn compare_mbti(results: &[serde_json::Value], lang: &str) -> serde_json::Value {
    let r1 = results.first();
    let r2 = results.get(1);

    let code1 = r1
        .and_then(|r| r.get("type_code"))
        .and_then(|c| c.as_str())
        .unwrap_or("");
    let code2 = r2
        .and_then(|r| r.get("type_code"))
        .and_then(|c| c.as_str())
        .unwrap_or("");

    let dichotomies = ["EI", "SN", "TF", "JP"];
    let mut matching = 0u32;
    let mut factor_details = serde_json::Map::new();

    let dichotomy_labels_en = ["E/I", "S/N", "T/F", "J/P"];
    let dichotomy_labels_es = ["E/I", "S/N", "T/F", "J/P"];
    let labels = if lang == "en" {
        &dichotomy_labels_en[..]
    } else {
        &dichotomy_labels_es[..]
    };

    for (i, d) in dichotomies.iter().enumerate() {
        let s1 = r1.map(|r| get_score(r, d)).unwrap_or(50.0);
        let s2 = r2.map(|r| get_score(r, d)).unwrap_or(50.0);

        let pole1 = pole_for(d, s1);
        let pole2 = pole_for(d, s2);
        let is_match = pole1 == pole2;
        if is_match {
            matching += 1;
        }

        let desc = if is_match {
            if lang == "en" {
                format!(
                    "{}: Both lean {} — aligned preference.",
                    labels[i], pole1
                )
            } else {
                format!(
                    "{}: Ambos inclinan hacia {} — preferencia alineada.",
                    labels[i], pole1
                )
            }
        } else {
            if lang == "en" {
                format!(
                    "{}: {} vs {} — complementary perspective.",
                    labels[i], pole1, pole2
                )
            } else {
                format!(
                    "{}: {} vs {} — perspectiva complementaria.",
                    labels[i], pole1, pole2
                )
            }
        };

        factor_details.insert(
            d.to_string(),
            serde_json::json!({
                "score1": (s1 * 10.0).round() / 10.0,
                "score2": (s2 * 10.0).round() / 10.0,
                "pole1": pole1,
                "pole2": pole2,
                "match": is_match,
                "description": desc,
            }),
        );
    }

    let (category, compatibility, summary) = match matching {
        4 => (
            "identical",
            90.0,
            if lang == "en" {
                "Identical types! You share the same MBTI type — strong mutual understanding."
            } else {
                "¡Tipos idénticos! Comparten el mismo tipo MBTI — fuerte entendimiento mutuo."
            },
        ),
        3 => (
            "similar",
            75.0,
            if lang == "en" {
                "Very similar types with only minor differences in one dichotomy."
            } else {
                "Tipos muy similares con solo diferencias menores en una dicotomía."
            },
        ),
        2 => (
            "complementary",
            60.0,
            if lang == "en" {
                "Complementary types — you balance each other across the dichotomies."
            } else {
                "Tipos complementarios — se equilibran mutuamente en las dicotomías."
            },
        ),
        _ => (
            "challenging",
            40.0,
            if lang == "en" {
                "Opposite types — your differences may create a challenging dynamic."
            } else {
                "Tipos opuestos — sus diferencias pueden crear una dinámica desafiante."
            },
        ),
    };

    serde_json::json!({
        "test_type": "mbti",
        "type_code1": code1,
        "type_code2": code2,
        "compatibility_score": compatibility,
        "match_category": category,
        "description": summary,
        "factors": factor_details,
    })
}

fn pole_for(dichotomy: &str, score: f64) -> &'static str {
    match dichotomy {
        "EI" => {
            if score < 50.0 {
                "E"
            } else {
                "I"
            }
        }
        "SN" => {
            if score < 50.0 {
                "S"
            } else {
                "N"
            }
        }
        "TF" => {
            if score < 50.0 {
                "T"
            } else {
                "F"
            }
        }
        "JP" => {
            if score < 50.0 {
                "J"
            } else {
                "P"
            }
        }
        _ => "",
    }
}

File: backend/app/compatibility.py
"""Personality type / profile compatibility comparisons.

Supports Big Five factor-level comparison and MBTI type matching
with bilingual (es/en) descriptions.
"""

from __future__ import annotations

from typing import Any

BIG_FIVE_FACTORS = ["O", "C", "E", "A", "N"]

FACTOR_LABELS: dict[str, dict[str, str]] = {
    "es": {
        "O": "Apertura a la Experiencia",
        "C": "Responsabilidad",
        "E": "Extraversión",
        "A": "Amabilidad",
        "N": "Neuroticismo",
    },
    "en": {
        "O": "Openness to Experience",
        "C": "Conscientiousness",
        "E": "Extraversion",
        "A": "Agreeableness",
        "N": "Neuroticism",
    },
}

# MBTI type compatibility mapping.
# Scores: identical=90, similar=75, complementary=60, challenging=40.
_MBTI_COMPAT: dict[str, dict[str, int]] = {
    # ENFJ
    "ENFJ": {"ENFJ": 90, "ENFP": 75, "INFJ": 75, "INFP": 60,
             "ENTJ": 60, "ENTP": 60, "INTJ": 60, "INTP": 40,
             "ESFJ": 75, "ESFP": 60, "ISFJ": 60, "ISFP": 40,
             "ESTJ": 40, "ESTP": 40, "ISTJ": 40, "ISTP": 40},
    # ENFP
    "ENFP": {"ENFP": 90, "ENFJ": 75, "INFP": 75, "INFJ": 60,
             "ENTP": 75, "ENTJ": 60, "INTJ": 60, "INTP": 60,
             "ESFP": 75, "ESFJ": 60, "ISFP": 60, "ISFJ": 40,
             "ESTP": 60, "ESTJ": 40, "ISTP": 40, "ISTJ": 40},
    # INFJ
    "INFJ": {"INFJ": 90, "INFP": 75, "ENFJ": 75, "ENFP": 60,
             "INTJ": 75, "INTP": 60, "ENTJ": 60, "ENTP": 40,
             "ISFJ": 60, "ISFP": 60, "ESFJ": 40, "ESFP": 40,
             "ISTJ": 40, "ISTP": 40, "ESTJ": 40, "ESTP": 40},
    # INFP
    "INFP": {"INFP": 90, "INFJ": 75, "ENFP": 75, "ENFJ": 60,
             "INTP": 75, "INTJ": 60, "ENTP": 60, "ENTJ": 40,
             "ISFP": 75, "ISFJ": 60, "ESFP": 60, "ESFJ": 40,
             "ISTP": 40, "ISTJ": 40, "ESTP": 40, "ESTJ": 40},
    # ENTJ
    "ENTJ": {"ENTJ": 90, "ENTP": 75, "INTJ": 75, "INTP": 60,
             "ENFJ": 60, "ENFP": 60, "INFJ": 60, "INFP": 40,
             "ESTJ": 75, "ESTP": 60, "ISTJ": 60, "ISTP": 40,
             "ESFJ": 40, "ESFP": 40, "ISFJ": 40, "ISFP": 40},
    # ENTP
    "ENTP": {"ENTP": 90, "ENTJ": 75, "INTP": 75, "INTJ": 60,
             "ENFP": 75, "ENFJ": 60, "INFJ": 60, "INFP": 60,
             "ESTP": 75, "ESTJ": 60, "ISTP": 60, "ISTJ": 40,
             "ESFP": 60, "ESFJ": 40, "ISFP": 40, "ISFJ": 40},
    # INTJ
    "INTJ": {"INTJ": 90, "INTP": 75, "ENTJ": 75, "ENTP": 60,
             "INFJ": 75, "INFP": 60, "ENFJ": 60, "ENFP": 40,
             "ISTJ": 60, "ISTP": 60, "ESTJ": 40, "ESTP": 40,
             "ISFJ": 40, "ISFP": 40, "ESFJ": 40, "ESFP": 40},
    # INTP
    "INTP": {"INTP": 90, "INTJ": 75, "ENTP": 75, "ENTJ": 60,
             "INFP": 75, "INFJ": 60, "ENFP": 60, "ENFJ": 40,
             "ISTP": 60, "ISTJ": 60, "ESTP": 40, "ESTJ": 40,
             "ISFP": 40, "ISFJ": 40, "ESFP": 40, "ESFJ": 40},
    # ESFJ
    "ESFJ": {"ESFJ": 90, "ESFP": 75, "ISFJ": 75, "ISFP": 60,
             "ENFJ": 75, "ENFP": 60, "INFJ": 60, "INFP": 40,
             "ESTJ": 60, "ESTP": 40, "ISTJ": 60, "ISTP": 40,
             "ENTJ": 40, "ENTP": 40, "INTJ": 40, "INTP": 40},
    # ESFP
    "ESFP": {"ESFP": 90, "ESFJ": 75, "ISFP": 75, "ISFJ": 60,
             "ENFP": 75, "ENFJ": 60, "INFP": 60, "INFJ": 40,
             "ESTP": 60, "ESTJ": 40, "ISTP": 60, "ISTJ": 40,
             "ENTP": 60, "ENTJ": 40, "INTP": 40, "INTJ": 40},
    # ISFJ
    "ISFJ": {"ISFJ": 90, "ISFP": 75, "ESFJ": 75, "ESFP": 60,
             "INFJ": 60, "INFP": 60, "ENFJ": 40, "ENFP": 40,
             "ISTJ": 60, "ISTP": 60, "ESTJ": 40, "ESTP": 40,
             "INTJ": 40, "INTP": 40, "ENTJ": 40, "ENTP": 40},
    # ISFP
    "ISFP": {"ISFP": 90, "ISFJ": 75, "ESFP": 75, "ESFJ": 60,
             "INFP": 75, "INFJ": 60, "ENFP": 60, "ENFJ": 40,
             "ISTP": 60, "ISTJ": 60, "ESTP": 40, "ESTJ": 40,
             "INTP": 40, "INTJ": 40, "ENTP": 40, "ENTJ": 40},
    # ESTJ
    "ESTJ": {"ESTJ": 90, "ESTP": 75, "ISTJ": 75, "ISTP": 60,
             "ENTJ": 75, "ENTP": 60, "INTJ": 60, "INTP": 40,
             "ESFJ": 60, "ESFP": 40, "ISFJ": 60, "ISFP": 40,
             "ENFJ": 40, "ENFP": 40, "INFJ": 40, "INFP": 40},
    # ESTP
    "ESTP": {"ESTP": 90, "ESTJ": 75, "ISTP": 75, "ISTJ": 60,
             "ENTP": 75, "ENTJ": 60, "INTP": 60, "INTJ": 40,
             "ESFP": 60, "ESFJ": 40, "ISFP": 60, "ISFJ": 40,
             "ENFP": 60, "ENFJ": 40, "INFP": 40, "INFJ": 40},
    # ISTJ
    "ISTJ": {"ISTJ": 90, "ISTP": 75, "ESTJ": 75, "ESTP": 60,
             "INTJ": 60, "INTP": 60, "ENTJ": 40, "ENTP": 40,
             "ISFJ": 60, "ISFP": 60, "ESFJ": 40, "ESFP": 40,
             "INFJ": 40, "INFP": 40, "ENFJ": 40, "ENFP": 40},
    # ISTP
    "ISTP": {"ISTP": 90, "ISTJ": 75, "ESTP": 75, "ESTJ": 60,
             "INTP": 60, "INTJ": 60, "ENTP": 40, "ENTJ": 40,
             "ISFP": 60, "ISFJ": 60, "ESFP": 40, "ESFJ": 40,
             "INFP": 40, "INFJ": 40, "ENFP": 40, "ENFJ": 40},
}

_MBTI_LABELS: dict[str, dict[str, str]] = {
    "es": {
        "identical": "Identical — máxima compatibilidad",
        "similar": "Similar — alta compatibilidad",
        "complementary": "Complementaria — buena compatibilidad",
        "challenging": "Desafiante — diferencias significativas",
    },
    "en": {
        "identical": "Identical — maximum compatibility",
        "similar": "Similar — high compatibility",
        "complementary": "Complementary — good compatibility",
        "challenging": "Challenging — significant differences",
    },
}

_DESCRIPTIONS: dict[str, dict[str, str]] = {
    "es": {
        "identical": "Ambos comparten el mismo tipo de personalidad. "
        "Se entienden con facilidad y tienen estilos de comunicación similares.",
        "similar": "Comparten tres de las cuatro letras del tipo MBTI. "
        "Existe una base sólida de entendimiento mutuo.",
        "complementary": "Comparten dos letras. "
        "Sus diferencias pueden fortalecer la relación al complementarse.",
        "challenging": "Comparten una o ninguna letra. "
        "Pueden enfrentar malentendidos, pero también oportunidades de crecimiento.",
    },
    "en": {
        "identical": "Both share the same personality type. "
        "They understand each other easily with similar communication styles.",
        "similar": "They share three of the four MBTI letters. "
        "A solid foundation for mutual understanding.",
        "complementary": "They share two letters. "
        "Differences can strengthen the relationship through complementarity.",
        "challenging": "They share one or no letters. "
        "May face misunderstandings, but also growth opportunities.",
    },
}


def _big_five_compatibility(
    result_a: dict[str, float],
    result_b: dict[str, float],
    lang: str,
) -> dict[str, Any]:
    """Calculate Big Five compatibility between two profiles."""
    factors: list[dict[str, Any]] = []
    total_diff = 0.0

    for factor in BIG_FIVE_FACTORS:
        a_val = result_a.get(factor, 50)
        b_val = result_b.get(factor, 50)
        diff = round(abs(a_val - b_val), 1)
        total_diff += diff
        label = FACTOR_LABELS.get(lang, FACTOR_LABELS["en"]).get(factor, factor)
        factors.append({"factor": factor, "label": label, "diff": diff,
                        "a": a_val, "b": b_val})

    # Compatibility score: 0 diff → 100, max possible diff per factor is 100,
    # over 5 factors = 500.  Invert so lower diff → higher score.
    max_diff = 500.0
    compatibility_score = round(max(0, 100 - (total_diff / max_diff) * 100), 1)

    if lang == "es":
        desc = (
            f"Compatibilidad Big Five: {compatibility_score}%. "
            f"Las diferencias en cada factor suman {total_diff} puntos."
        )
    else:
        desc = (
            f"Big Five compatibility: {compatibility_score}%. "
            f"Factor differences total {total_diff} points."
        )

    return {"compatibility_score": compatibility_score, "description": desc,
            "factors": factors}


def _mbti_compatibility(
    type_a: str,
    type_b: str,
    lang: str,
) -> dict[str, Any]:
    """Calculate MBTI type compatibility."""
    type_a = type_a.upper().strip()
    type_b = type_b.upper().strip()

    row = _MBTI_COMPAT.get(type_a, {})
    score = row.get(type_b, 50)

    if score >= 85:
        level = "identical"
    elif score >= 70:
        level = "similar"
    elif score >= 50:
        level = "complementary"
    else:
        level = "challenging"

    labels = _MBTI_LABELS.get(lang, _MBTI_LABELS["en"])
    descs = _DESCRIPTIONS.get(lang, _DESCRIPTIONS["en"])

    description = f"{labels[level]}. {descs[level]}"

    return {
        "compatibility_score": score,
        "description": description,
        "factors": [
            {"label": "MBTI Type Match", "level": level,
             "type_a": type_a, "type_b": type_b},
        ],
    }


def compare_results(
    result_a: dict[str, Any],
    result_b: dict[str, Any],
    test_type: str,
    lang: str = "es",
) -> dict[str, Any]:
    """Compare two personality profiles and return compatibility info.

    Parameters
    ----------
    result_a, result_b:
        Full result dicts as returned by the scoring modules.
        Each should contain at least a ``scores`` sub-dict (Big Five)
        or a ``type`` key (MBTI).
    test_type:
        One of ``"big_five"``, ``"mbti"``.
    lang:
        Language code for descriptions (``"es"`` or ``"en"``).

    Returns
    -------
    dict with keys ``compatibility_score``, ``description``, ``factors``.
    """
    if test_type == "big_five":
        scores_a = result_a.get("scores", result_a)
        scores_b = result_b.get("scores", result_b)
        return _big_five_compatibility(scores_a, scores_b, lang)

    if test_type == "mbti":
        type_a = result_a.get("type", "INFJ")
        type_b = result_b.get("type", "INFJ")
        return _mbti_compatibility(type_a, type_b, lang)

    # Fallback: unknown test type
    labels = FACTOR_LABELS.get(lang, FACTOR_LABELS["en"])
    if lang == "es":
        desc = f"La compatibilidad para {test_type} no está implementada."
    else:
        desc = f"Compatibility for {test_type} is not implemented yet."
    return {"compatibility_score": 50, "description": desc, "factors": [
        {"label": labels.get("O", "Unknown"), "diff": 0},
    ]}
